### PR TITLE
Election scheduler

### PR DESCRIPTION
**Pull request by clemahieu**
This change replaces PoW-difficulty based prioritization with balance * time_since_use based prioritization. The scheduler currently manages two queues. A priority queue where live traffic is managed and scheduled according to vacancy in the active_transactions container and prioritized by the prioritization container. A manual queue where requests that have come through RPCs are enqueued and bypass prioritization.

### CODE REVIEW

**Review by theohax — DISMISSED** [review]

File: nano/core_test/election_scheduler.cpp [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+	ASSERT_NE (nullptr, system.nodes[0]->active.election (send1->qualified_root ()));
+}
+
+TEST (election_scheduler, no_vacancy)
+{
+	nano::system system;
+	nano::node_config config{ nano::get_available_port (), system.logging };
+	config.active_elections_size = 1;
+	auto & node = *system.add_node (config);
+	;
```
Comment (SergiySW) [inline nano/core_test/election_scheduler.cpp]: Seems not required

File: nano/core_test/election_scheduler.cpp [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+				   .work (*system.work.generate (key.pub))
+				   .build_shared ();
+	ASSERT_EQ (nano::process_result::progress, node.process (*send).code);
+	ASSERT_EQ (nano::process_result::progress, node.process (*receive).code);
+	node.block_confirm (send);
+	auto election1 = node.active.election (send->qualified_root ());
+	election1->force_confirm ();
+	node.block_confirm (receive);
+	auto election2 = node.active.election (receive->qualified_root ());
+	election2->force_confirm ();
```
Comment (SergiySW) [inline nano/core_test/election_scheduler.cpp]: Probably can be replaces with `nano::blocks_confirm (node, { send, receive }, true);`

File: nano/node/election_scheduler.cpp [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+	return priority.empty () && manual_queue.empty ();
+}
+
+size_t nano::election_scheduler::priority_queue_size () const
+{
+	return priority.size ();
+}
+
+void nano::election_scheduler::run ()
+{
```
Comment (SergiySW) [inline nano/node/election_scheduler.cpp:88]: Probably better to name thread as well?

File: nano/node/election_scheduler.cpp [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+}
+
+void nano::election_scheduler::manual (std::shared_ptr<nano::block> const & block_a, boost::optional<nano::uint128_t> const & previous_balance_a, nano::election_behavior election_behavior_a, std::function<void (std::shared_ptr<nano::block> const &)> const & confirmation_action_a)
+{
+	nano::lock_guard<nano::mutex> lock{ mutex };
+	manual_queue.push_back (std::make_tuple (block_a, previous_balance_a, election_behavior_a, confirmation_action_a));
+	observe ();
+}
+
+void nano::election_scheduler::activate (nano::account const & account_a, nano::transaction const & transaction)
```
Comment (SergiySW) [inline nano/node/election_scheduler.cpp:24]: `transaction` -> `transaction_a` ?

**Review by SergiySW — DISMISSED** [review]

**Review by SergiySW — APPROVED** [review]

**Pull request merged**